### PR TITLE
refactor: use os_pipe

### LINKED ISSUE #2: Share non-process stdin across multiple commands
stdin when not from the process' stdin needs to be shared across multiple commands. Ex. `echo "abcdefghijklmnop" | (deno eval "const b = new Uint8Array(10);Deno.stdin.readSync(b);console.log(b)" && deno eval "const b = new Uint8Array(10);Deno.stdin.readSync(b);console.log(b)")`

I'm not exactly sure how to do this, but will look into it later as it's an edge case that's not blocking.

Edit: Looks like https://crates.io/crates/os_pipe is already used in deno.

### PR DESCRIPTION
**Pull request by dsherret**
Massive improvement to the code by using this. Didn't know it existed.

https://crates.io/crates/os_pipe

This also will resolve #2, but I will open up a separate PR testing that.

### CODE REVIEW

**Comment by dsherret** [comment]
@bartlomieju I'm pretty confident in this one so going to merge it. I will open up a separate PR tomorrow to add tests for #2 and I can address any feedback in that PR if you see anything that should be resolved from these changes.

File: src/shell.rs [hunk 1]
```diff
@@ -92,434 +75,310 @@ enum AsyncCommandBehavior {
...
-  let step = spawn_sequence(sequence, state, stdin, environment).await;
-  step
-    .wait_with_stdout(move |stdout| async {
-      stdout.pipe_to_sender(sender).await;
-    })
-    .await
-}
-
-async fn start_command(
+fn start_command(
```
Comment (bartlomieju) [inline src/shell.rs:273]: Why return future instead of making the function `async`? Is it still caused by need to recursive calls?

**Review by bartlomieju — COMMENTED** [review]
I don't see any handling for stderr, is that on purpose?

File: src/shell.rs (hunk 1 above)
Comment (dsherret) [reply src/shell.rs:273]: Nope, no need for this. Will do this update in the next pr. Thanks!

**Comment by dsherret** [comment]
> I don't see any handling for stderr, is that on purpose?

Right now it just goes directly to stderr because there's no re-direct support. I'm going to add a pipe for stderr very soon.

**Pull request merged**